tools/c7n_gcp/c7n_gcp/filters/metrics.py

```diff
@@ -0,0 +1,181 @@
+# SPDX-License-Identifier: Apache-2.0
+"""
+Monitoring Metrics suppport for resources
+"""
+from datetime import datetime, timedelta
+import pytz
+
+import jmespath
+
+from c7n.filters.core import Filter, OPERATORS
+from c7n.utils import local_session, type_schema
+
+from c7n_gcp.provider import resources as gcp_resources
+
+REDUCERS = [
+    'REDUCE_NONE',
+    'REDUCE_MEAN',
+    'REDUCE_MIN',
+    'REDUCE_MAX',
+    'REDUCE_MEAN',
+    'REDUCE_SUM',
+    'REDUCE_STDDEV',
+    'REDUCE_COUNT',
+    'REDUCE_COUNT_TRUE',
+    'REDUCE_COUNT_FALSE',
+    'REDUCE_FRACTION_TRUE',
+    'REDUCE_PERCENTILE_99',
+    'REDUCE_PERCENTILE_95',
+    'REDUCE_PERCENTILE_50',
+    'REDUCE_PERCENTILE_05']
+
+ALIGNERS = [
+    'ALIGN_NONE',
+    'ALIGN_DELTA',
+    'ALIGN_RATE',
+    'ALIGN_INTERPOLATE',
+    'ALIGN_MIN',
+    'ALIGN_MAX',
+    'ALIGN_MEAN',
+    'ALIGN_COUNT',
+    'ALIGN_SUM',
+    'REDUCE_COUNT_FALSE',
+    'ALIGN_STDDEV',
+    'ALIGN_COUNT_TRUE',
+    'ALIGN_COUNT_FALSE',
+    'ALIGN_FRACTION_TRUE',
+    'ALIGN_PERCENTILE_99',
+    'ALIGN_PERCENTILE_95',
+    'ALIGN_PERCENTILE_50',
+    'ALIGN_PERCENTILE_05',
+    'ALIGN_PERCENT_CHANG']
+
+
+class GCPMetricsFilter(Filter):
+    """Supports metrics filters on resources.
+
+    All resources that have cloud watch metrics are supported.
+
+    Docs on cloud watch metrics
+
+    - Google Supported Metrics
+      https://cloud.google.com/monitoring/api/metrics_gcp
+
+    - Custom Metrics
+      https://cloud.google.com/monitoring/api/v3/metric-model#intro-custom-metrics
+
+    .. code-block:: yaml
+
+      - name: firewall-hit-count
+        resource: gcp.firewall
+        filters:
+        - type: metrics
+          name: firewallinsights.googleapis.com/subnet/firewall_hit_count
+          metric-key: metric.labels.firewall_name
+          resource-key: name
+          aligner: ALIGN_COUNT
+          days: 14
+          value: 1
+          op: greater-than
+    """
+
+    schema = type_schema(
+        'metrics',
+        **{'name': {'type': 'string'},
+          'resource-key': {'type': 'string'},
+          'metric-key': {'type': 'string'},
+          'group-by-fields': {'type': 'array', 'items': {'type': 'string'}},
+          'days': {'type': 'number'},
+          'op': {'type': 'string', 'enum': list(OPERATORS.keys())},
+          'reducer': {'type': 'string', 'enum': REDUCERS},
+          'aligner': {'type': 'string', 'enum': ALIGNERS},
+          'value': {'type': 'number'},
+          'filter': {'type': 'string'},
+          'period': {'type': 'number'},
+          'missing-value': {'type': 'number'},
+          'required': ('value', 'name', 'op')})
+    permissions = ("monitoring.timeSeries.list",)
+
+    def process(self, resources, event=None):
+        days = self.data.get('days', 14)
+        duration = timedelta(days)
+
+        self.metric = self.data['name']
+        self.resource_key = self.data['resource-key']
+        self.metric_key = self.data['metric-key']
```

Comment:
wondering if we should also add a default here to the resource type metadata at least for name based queries.

Reply:
i was  looking into this, but there does not seem to be any standard. in some it is metric.labels.instance_name or metric.labels.firewall_name and others i found resource.labels.function_name, so would most likely have to add a field to all of the gcp resource types.

Reply:
fair, we can merge as is, but i think it would help usability if we could annotate the type metadata, i checked about a half dozen, and it was generally same as name field, but understood its not consistent.

Reply:
thats a good point. certainly easier for users to not have to lookup and add that field in. I will go ahead and annotate resources with a metric_key.